pyfr/backends/base/makoutil.py

```diff
@@ -98,47 +99,73 @@ def macro(context, name, params, externs=''):
     params = [p.strip() for p in params.split(',')]
     externs = [e.strip() for e in externs.split(',')] if externs else []
 
-    # Ensure no invalid characters in params/extern variables
+    # Extract Python data parameters from caller signature
+    sig = inspect.signature(context['caller'].body).parameters
+    dparams = [p.name for p in sig.values()]
+
+    # Validate all parameter names
     for p in it.chain(params, externs):
         if not re.match(r'[A-Za-z_]\w*$', p):
-            raise ValueError(f'Invalid param "{p}" in macro "{name}"')
-
-    # Capture the function body
-    body = capture(context, context['caller'].body)
+            raise ValueError(f'Invalid param "{p}" in dmacro "{name}"')
 
-    # Identify any local variable declarations
-    lvars = _locals(body)
-
-    # Suffix these variables by a '_'
-    if lvars:
-        body = re.sub(r'\b({0})\b'.format('|'.join(lvars)), r'\1_', body)
-
-    # Save
-    context['_macros'][name] = (params, externs, body)
+    # Store for later capture
+    context['_macros'][name] = (
+        params,
+        externs,
+        dparams,
+        context['caller'].body
+    )
 
     return ''
 
 
 def expand(context, name, /, *args, **kwargs):
-    # Get the macro parameter list and the body
-    mparams, mexterns, body = context['_macros'][name]
 
-    # Ensure an appropriate number of arguments have been passed
-    if len(mparams) != len(args) + len(kwargs):
-        raise ValueError(f'Inconsistent macro parameter list in {name}')
-
-    # Parse the parameter list
-    params = dict(zip(mparams, args))
+    macro_info = context['_macros'][name]
+    mparams, mexterns, mdparams, macro_callable = macro_info
+
+    # Validate argument count
+    # Params can use args or kwargs; dparams must be positional
+    nparams_kw = len(kwargs)
+    nparams_pos = len(mparams) - nparams_kw
+    ndparams_pos = len(args) - nparams_pos
+
+    # Ensure correct counts
+    if nparams_pos + nparams_kw != len(mparams):
+        emps = f'Expected {len(mparams)} parameters in {name}'
```

Comment:
Move into the call (but break with a suitable newline; see some other examples).